### PR TITLE
RealtimeSegmentConverter was using incorrect schema

### PR DESCRIPTION
**Pull request by gortiz**
`RealtimeSegmentConverter` cannot use the realtime segment because it contains some virtual columns. For example the `$segmentName` and `$docId` will be different in the sealed segment. Therefore `RealtimeSegmentConverter` copies the schema, removing the virtual columns in the process.

That copy was done manually in `RealtimeSegmentConverter` and was a partial copy. For example, the generated schema doesn't keep the schema name. By chance the fact that this copy was partial didn't affect the sealing process. But when https://github.com/apache/pinot/pull/11960 was added the partial copy in `RealtimeSegmentConverter` had an important side effect: column based null handling was lost.

That means that the realtime segment contains null columns, but once it is sealed these vectors are ignored.

This PR fixes that issue and adds some regression tests, but given Schema is mutable it is very difficult to verify that there are no more incorrect copies in the code. A refactor of the Schema class to make it more secure is needed.

### CODE REVIEW

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 1]
```diff
@@ -526,6 +544,14 @@ public SchemaBuilder() {
       _schema = new Schema();
     }
 
+    public static SchemaBuilder copyWithoutFields(Schema other) {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:547]: Is this needed?

**Review by Jackie-Jiang — APPROVED** [review]
Good catch!

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java [hunk 2]
```diff
@@ -352,6 +352,24 @@ public TreeSet<String> getPhysicalColumnNames() {
     return physicalColumnNames;
   }
 
+  /**
+   * Returns a new schema containing only physical columns.
+   *
+   * All properties but the fields are the same.
+   * All field attributes are a shallow copy without the virtual column.
+   */
+  public Schema withoutVirtualColumns() {
```
Comment (yashmayya) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:361]: While there might not be any current use case for it, I think it'd be good to also copy over the primary key columns to the new schema?

**Review by yashmayya — APPROVED** [review]
Thanks @gortiz, good catch! I just had one minor comment.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 2 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:361]: You are right. Added (and added some tests)

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java (hunk 1 above)
Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java:547]: No. I was planning to use this method to fix the clone method, but it wasn't trivial (and it was not related to this PR).

I'm removing this method

**Pull request merged**